### PR TITLE
Batch inference feature branch

### PR DESCRIPTION
**Pull request by amogkam**
Feature branch for development of https://github.com/ray-project/enhancements/pull/25

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: predict.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        self.use_gpu = use_gpu
+        super().__init__(preprocessor)
+
+    def _predict_pandas(self, data: pd.DataFrame, **kwargs) -> pd.DataFrame:
+        return data * self.factor
+
+
+predictor = DummyPredictor(factor=2)
+ds = ray.data.range_table(4)
+ds = ds.map_batches(predictor, compute=ActorPoolStrategy(4, 4))
```
Comment (ericl) [inline predict.py:38]:
```suggestion
ds = ds.map_batches(predictor, compute=ActorPoolStrategy(size=4))
```

Once we merge https://github.com/ray-project/ray/pull/33771

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -622,6 +624,46 @@ def map_batches(
...
+
+            checkpoint = Checkpoint.from_dict({"predictor": fn})
+
+            batch_predictor = BatchPredictor.from_checkpoint(
+                checkpoint, predictor_cls=_WrapperPredictor
+            )
+
+            assert isinstance(compute, ActorPoolStrategy)
+            kwargs = fn_kwargs if fn_kwargs else {}
+            return batch_predictor.predict(
```
Comment (ericl) [inline python/ray/data/dataset.py:656]: Let's get rid of this dependency on BatchPredictor? We want to deprecate that class.

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -622,6 +624,46 @@ def map_batches(
...
+                def predict(self, data, **kwargs):
+                    return self.base_predictor.predict(data, **kwargs)
+
+                def _predict_pandas(self, data, **kwargs):
+                    return self.base_predictor._predict_pandas(data, **kwargs)
+
+                def _predict_numpy(self, data, **kwargs):
+                    return self.base_predictor._predict_numpy(data, **kwargs)
+
+            checkpoint = Checkpoint.from_dict({"predictor": fn})
```
Comment (ericl) [inline python/ray/data/dataset.py:648]: This isn't lazy loaded right.

**Comment by amogkam** [comment]
@ericl to clarify, this is an initial start so that the APIs can immediately be used for CUJ. This is not the actual implementation that we want to get merged in master.

**Comment by amogkam** [comment]
merging into a feature branch

**Pull request merged**